### PR TITLE
feat: adding denylist package (closes #68)

### LINKED ISSUE #68: denylist package
## Motivation

> Found an issue when a given worker does 2+ interactions with same worker binding during the same request. It weirdly reuses same Request instance (updated with new request), which consequently makes our logger/sentry not properly reporting the request.url + request.method (they are overwritten…). If we just clone the request to provide Logger and Sentry, then all follow up requests come with same first request details…

## Plan

Create denylist package and integrate it:

- [x] create denylist package 
- [x] move http route from `cid-verifier`
- [ ] move cron job from `edge-gateway` https://github.com/web3-storage/reads/blob/main/.github/workflows/cron-denylist.yml
- [x] update edge-gateway calls and worker bindings

### PR DESCRIPTION
**Pull request by jsdevel**

### CODE REVIEW

File: packages/denylist/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# denylist
+
+> The `denylist` package serves up data from badbits denylists.
```
Comment (vasco-santos) [inline packages/denylist/README.md:3]:
```suggestion
> The `denylist` package serves up data from dotstorage denylists (includes [badbits](https://badbits.dwebops.pub/)).
```

File: packages/denylist/package.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "denylist",
+  "version": "1.0.0",
+  "description": "Provide badbits denylist data.",
```
Comment (vasco-santos) [inline packages/denylist/package.json:4]:
```suggestion
  "description": "Denylist data for dotstorage reads projects",
```

File: packages/denylist/scripts/build.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import git from 'git-rev-sync'
+import Sentry from '@sentry/cli'
+
+const __dirname = path.dirname(fileURLToPath(import.meta.url))
+const pkg = JSON.parse(
+  fs.readFileSync(path.join(__dirname, '..', 'package.json'), 'utf8')
+)
+
+export async function buildCmd (opts) {
+  const sentryRelease = `cid-verifier@${pkg.version}-${opts.env}+${git.short(
```
Comment (vasco-santos) [inline packages/denylist/scripts/build.js:14]:
```suggestion
  const sentryRelease = `denylist-api@${pkg.version}-${opts.env}+${git.short(
```

File: packages/denylist/scripts/build.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    minify: opts.env !== 'dev',
+    sourcemap: 'external'
+  })
+
+  // Sentry release and sourcemap upload
+  if (process.env.SENTRY_UPLOAD === 'true') {
+    const cli = new Sentry(undefined, {
+      authToken: process.env.SENTRY_TOKEN,
+      org: 'protocol-labs-it',
+      project: 'cid-verifier',
```
Comment (vasco-santos) [inline packages/denylist/scripts/build.js:41]:
```suggestion
      project: 'denylist-api',
```

File: packages/denylist/scripts/cli.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env node
+
+import sade from 'sade'
+
+import { buildCmd } from './build.js'
+
+const env = process.env.ENV || 'dev'
+const prog = sade('cid-verifier')
```
Comment (vasco-santos) [inline packages/denylist/scripts/cli.js:8]:
```suggestion
const prog = sade('denylist')
```

File: packages/denylist/src/denylist.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import { normalizeCid } from './utils/cid'
+import { getFromDenyList } from './utils/denylist'
+import { JSONResponse } from '@web3-storage/worker-utils/response'
+
+/**
+ * @param {Array<string>} params
+ * @param {(params: Array<string>, request: Request, env: import('./env').Env) => Promise<Response>} fn
+ * @returns {import('itty-router').RouteHandler<Request>}
+ */
+function withRequiredQueryParams (params, fn) {
```
Comment (vasco-santos) [inline packages/denylist/src/denylist.js:12]: Thoughts on making this API just?

`GET /:cid` ? So, no query parameters. It feels wait more conventional REST to me

File: packages/denylist/src/denylist.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  /**
+   * Returns badbits denylist results.
+   */
+  async function (params, request, env) {
+    const [cid] = params
+
+    const denyListResource = await getFromDenyList(cid, env)
+    if (denyListResource) {
+      try {
+        return new JSONResponse(JSON.parse(denyListResource), { status: 200 })
```
Comment (vasco-santos) [inline packages/denylist/src/denylist.js:49]: I agree with making this 200 + 404 👍🏼  and we pass what we know from the CID, letting consumers take needed actions

File: packages/denylist/src/errors.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  constructor (message = 'Service Unavailable') {
+    const status = 503
+    super(message)
+    this.name = 'ServiceUnavailableError'
+    this.status = status
+    this.code = ServiceUnavailableError.CODE
+    this.contentType = 'text/html'
+  }
+}
+ServiceUnavailableError.CODE = 'ERROR_SERVICE_UNAVAILABLE'
```
Comment (vasco-santos) [inline packages/denylist/src/errors.js:14]: no need for this error

File: packages/denylist/src/index.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+const router = Router()
+
+// https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent
+/** @typedef {ExecutionContext} Ctx */
+
+router
+  .all('*', envAll)
+  .get('/version', withCorsHeaders(versionGet))
+  .get('/', withCorsHeaders(denylistGet))
```
Comment (vasco-santos) [inline packages/denylist/src/index.js:20]: Per feedback above, let's do:

```suggestion
  .get('/:cid', withCorsHeaders(denylistGet))
```

File: packages/denylist/test/mocks/package.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (vasco-santos) [inline packages/denylist/test/mocks/package.json:1]: We don't need this file

File: packages/denylist/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "outDir": "dist",
+    "noEmit": true,
+    "lib": ["ESNext"],
+    "types": ["@cloudflare/workers-types"]
+  },
+  "include": ["src", "test", "package.json"],
+  "exclude": ["**/node_modules/**", "test/mocks/**"]
```
Comment (vasco-santos) [inline packages/denylist/tsconfig.json:10]:
```suggestion
  "exclude": ["**/node_modules/**"]
```

File: packages/denylist/wrangler.toml [hunk 12]
```diff
@@ -0,0 +1,54 @@
+# dotstorage cid verifier wrangler config.
```
Comment (vasco-santos) [inline packages/denylist/wrangler.toml:1]:
```suggestion
# dotstorage denylist wrangler config.
```

File: packages/denylist/wrangler.toml [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+name = "dotstorage-denylist"
+main = "./dist/worker.js"
+compatibility_date = "2022-07-01"
```
Comment (vasco-santos) [inline packages/denylist/wrangler.toml:4]:
```suggestion
compatibility_date = "2022-09-27"
```

**Review by vasco-santos — CHANGES_REQUESTED** [review]
Mostly there, given feedback should be enough :)

File: packages/denylist/src/index.js (hunk 9 above)
Comment (vasco-santos) [reply packages/denylist/src/index.js:20]:
```suggestion
```

So this line should now go away

File: packages/denylist/src/denylist.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+import { getFromDenyList } from './utils/denylist'
+import { JSONResponse } from '@web3-storage/worker-utils/response'
+
+/**
+ * Returns badbits denylist results.
+ * @param {Request} request
+ * @param {import('./env').Env} env
+ */
+export const denylistGet = async function (request, env) {
+  const cid = (new URL(request.url)).pathname?.substring(1)
```
Comment (vasco-santos) [inline packages/denylist/src/denylist.js:13]: You should be able to do here, dropping the query param kind of route that I commented in index.js file

```suggestion
  const cid = request.params.cid
```

**Review by vasco-santos — COMMENTED** [review]
Last 2 comments and we are ready to go. Good work @jsdevel

File: packages/denylist/src/index.js (hunk 9 above)
Comment (jsdevel) [reply packages/denylist/src/index.js:20]: i kept this only to give API users a more meaningful error message if they request `/` without providing a cid.  but i guess from an HTTP perspective, if they request `/`, that's technically a different resource.  so 404 would be acceptable.

File: packages/denylist/src/denylist.js (hunk 14 above)
Comment (vasco-santos) [reply packages/denylist/src/denylist.js:13]: The type request is not Web Request, but itty router request https://github.com/kwhitley/itty-router#example

So, types need to be updated to receive `@param {import('itty-router')} request` instead

File: packages/denylist/src/index.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    try {
+      const res = await router.handle(req, env, ctx)
+      env.log.timeEnd('request')
+      return env.log.end(res)
+    } catch (/** @type {any} */ error) {
+      if (env.log) {
+        env.log.timeEnd('request')
+        return env.log.end(serverError(error, req, env))
+      }
+      return serverError(error, req, env)
```
Comment (vasco-santos) [inline packages/denylist/src/index.js:50]: I submitted a PR to change this that got merged https://github.com/web3-storage/reads/pull/86

but it is not here. Mind have a look @jsdevel ?

File: packages/denylist/src/env.js [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    token: env.LOKI_TOKEN,
+    debug: Boolean(env.DEBUG),
+    version: env.VERSION,
+    commit: env.COMMITHASH,
+    branch: env.BRANCH,
+    worker: 'cid-verifier',
+    env: env.ENV,
+    sentry: env.sentry
+  })
+  env.log.time('request')
```
Comment (vasco-santos) [inline packages/denylist/src/env.js:38]: same as other comment

**Review by vasco-santos — CHANGES_REQUESTED** [review]
@jsdevel this looks good to me. We just need to integrate the PR mentioned in comments and is ready to go

File: packages/denylist/src/index.js (hunk 15 above)
Comment (jsdevel) [reply packages/denylist/src/index.js]: done.  i never pull down my feature branches, so i likely overwrote your changes.

File: packages/denylist/src/env.js (hunk 16 above)
Comment (jsdevel) [reply packages/denylist/src/env.js]: done

**Review by vasco-santos — APPROVED** [review]
LGTM

**Pull request merged**